### PR TITLE
Merge highlevel3 into integration

### PR DESCRIPTION
**Pull request by xaterz**
What has been done in this branch:
* Implemented the 'Modules that are taken prior to other modules' page (shows all pairs)
* Implemented search function for 'number of students who took a specified module A prior to a specified module B in a target AY-Sem'

### CODE REVIEW

File: components/model.py [hunk 1]
```diff
@@ -656,3 +656,96 @@ def get_all_mods_taken_together():
     DB_CURSOR.execute(sql_command)
```
Comment (nlzz22) [inline components/model.py]: please pylint this file

File: templates/modulesTakenPriorToOthers.html [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+$var title:View Modules Taken Prior to Others
+
+$if modulePairs is not None:
+	<div class="container-fluid">
+		<h1 class="text-center"><b>Modules that are taken prior to other modules</b></h1>
+		<p class="text-center">Shows all pairs of modules where module A is taken prior to module B</b></p>
+		<br>
+		<div class="modules-taken-prior-search-form">
+			<form action="/moduleTakenPriorToOthers" method="POST">
```
Comment (tgqiang) [inline templates/modulesTakenPriorToOthers.html]: Validation is needed for both module code inputs.

You want to leave it to UI department to fix or you want to implement it?

File: templates/modulesTakenPriorToOthers.html [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+					$ percentage = studentPriorCount*1.0/studentsInModuleB*100
+				<tr>
+					<td style="text-align: right;">Percentage of students who took $moduleA 
+					before $moduleB:</td>
+					<td style="padding-left: 30px;"><b>$percentage%</b></td>
+				</tr>
+			</tbody>
+		</table>
+
+		<form style="margin-top: 25px;" action="/moduleTakenPriorToOthers" method="post">
```
Comment (tgqiang) [inline templates/modulesTakenPriorToOthers.html]: Why does this <code>Go Back</code> navigation use the POST method? A GET method makes more sense.

File: templates/modulesTakenPriorToOthers.html [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+$var title:View Modules Taken Prior to Others
+
+$if modulePairs is not None:
+	<div class="container-fluid">
+		<h1 class="text-center"><b>Modules that are taken prior to other modules</b></h1>
+		<p class="text-center">Shows all pairs of modules where module A is taken prior to module B</b></p>
+		<br>
+		<div class="modules-taken-prior-search-form">
+			<form action="/moduleTakenPriorToOthers" method="POST">
+				<p>Tell me the number of students who have taken</p>
```
Comment (tgqiang) [inline templates/modulesTakenPriorToOthers.html]: "Tell me the number of students who have taken **A** before they are *going to* take **B** in **a specified AY-Semester**"

I tried injecting a <code>studentplan</code> row with the following: ['D5000000A', true, 'CS1020', 'AY 16/17 Sem 2']
(meaning student D5000000A has already taken CS1020 in AY 16/17 Sem 2)

This particular student was also taken into account for the query "Find students who have taken CS1010 before they are going to take CS1020 in AY 16/17 Sem 2".

Due to the **going to** clause, that student should not have been considered seeing that he has taken CS1020 in AY 16/17 Sem 2 **already**.

Thus, I wish to double-check whether students who HAVE TAKEN the module already should or should not be considered in this query.

Comment (tgqiang) [inline templates/modulesTakenPriorToOthers.html]: The implication of the above being either the UI changes, or the back-end changes.

**Comment by xaterz** [comment]
Most of the issues raised have been fixed. Please review the commits from 4296b0e onwards.

My reply to some of the issues:

* I don't do test cases for UI. After this is approved, @tgqiang or @a0129998 please write the UI test cases for this feature, thanks.
* The reason why I always use POST method for navigation, is because if the GET method is used, there will be a trailing question mark in the URL, which doesn't look very nice.

image

* But since this has been raised, I have changed the Go Back button to use javascript redirect instead. @tgqiang if you want to totally avoid using POST for navigation, you need to change for ALL the navigation buttons currently in our app.
* The 'going to take' issue is just a matter of phrasing. I have changed it to 'taking' instead. The ambiguity was actually explained in my code comment if you haven't notice:

image

**Comment by nlzz22** [comment]
Pull request is approved: Issues raised previously have been addressed, all codes are linted and all test cases passed.

**Pull request merged**